src/App.jsx

```diff
@@ -1,20 +1,83 @@
 import './App.css';
 import TodoList from './TodoList';
 import AddTodoForm from './AddTodoForm';
-import { useState, useEffect } from 'react';
+import { useState, useEffect, useCallback } from 'react';
+import axios from 'axios';
+import {
+  BrowserRouter,
+  Routes,
+  Route
+} from "react-router-dom";
 
-const useSemiPersistentState = () => {
-  const [value, setValue] = useState(JSON.parse(localStorage.getItem('savedTodoList')) || []);
-  useEffect(() => {
-    localStorage.setItem('savedTodoList', JSON.stringify(value));
-  }, [value]);
-  return [value, setValue];
-};
+const url = `https://api.airtable.com/v0/${import.meta.env.VITE_AIRTABLE_BASE_ID}/${import.meta.env.VITE_TABLE_NAME}`;
+const bearerToken = `Bearer ${import.meta.env.VITE_AIRTABLE_API_TOKEN}`;
 
 const App = () => {
-  const [todoList, setTodoList] = useSemiPersistentState();
-  const addTodo = (newTodo) => {
-    setTodoList([newTodo, ...todoList])
+  const [todoList, setTodoList] = useState([]);
+  const [isLoading, setIsLoading] = useState(true);
+  const [isAdding, setIsAdding] = useState(false);
+
+  // useCallBack for pessimistic rendering, to make sure the displayed data matches the db
+  const callbackFetchData = useCallback(async () => {
+    const options = {
+      headers: { Authorization: bearerToken },
+    };
+    try {
+      const response = await axios.get(url, options);
+      if (response.status !== 200) throw new Error(`${response.status}`);
+
+      const data = response.data.records;
+      const todos = data.map(todo => todo = { title: todo.fields.title, id: todo.id });
+      return todos;
+    } catch (error) {
+      console.log(error);
+    }
+  }, [isLoading, isAdding]);
+
+  useEffect(() => {
+    if (isLoading) {
+      new Promise((resolve, reject) => { setTimeout(() => resolve(callbackFetchData()), 2000) })
+        .then(result => {
+          setTodoList(result || []);
+          setIsLoading(false);
+        });
+    }
+
+    if (isAdding) {
+      new Promise((resolve, reject) => resolve(callbackFetchData()))
+        .then(result => {
+          setTodoList(result || []);
+          setIsAdding(false);
+        });
+    }
+  }, [callbackFetchData]);
+
+  useEffect(() => {
+    if (!isLoading) {
+      localStorage.setItem('savedTodoList', JSON.stringify(todoList));
+    }
+  }, [todoList]);
+
+  const addTodo = async (newTodo) => {
+    let todo = JSON.stringify({ "fields": newTodo });
+    const options = {
+      headers: {
+        "Content-Type": "application/json",
+        Authorization: bearerToken
+      },
+    };
+
+    try {
+      const response = await axios.post(url, todo, options);
+      if (response.status !== 200) throw new Error(`${response.status}`);
+
+      const { id: todoId, fields: { title: todoTitle } } = response.data;
+      todo = { id: todoId, title: todoTitle };
+      setIsAdding(true);
+      setTodoList([todo, ...todoList]);
```

Comment:
So your todoList will change twice here (if they are not sorted the same way). Once with the fetch request started by setting `isAdding` to true and again with your `setTodoList` here